src/Nest/Document/Multiple/MultiGet/Response/MultiGetResponse.cs

```diff
@@ -27,12 +27,28 @@ public class MultiGetResponse : ResponseBase
 			return multiHit?.Fields ?? FieldValues.Empty;
 		}
 
+		/// <summary>
+		/// Retrieves the hits for each distinct id.
+		/// </summary>
+		/// <param name="ids">The ids to retrieve source for</param>
+		/// <typeparam name="T">The document type for the hits to return</typeparam>
+		/// <returns>An IEnumerable{T} of hits</returns>
 		public IEnumerable<IMultiGetHit<T>> GetMany<T>(IEnumerable<string> ids) where T : class
 		{
-			var docs = Hits.OfType<IMultiGetHit<T>>();
-			return from d in docs
-				join id in ids on d.Id equals id
-				select d;
+			HashSet<string> seenIndices = null;
+			foreach (var id in ids.Distinct())
+			{
+				if (seenIndices == null)
```

Comment:
I wonder if we could rewrite this using `GroupBy()` instead. I understand how this initialization works but will probably do a double take every time I will read it in the future

Reply:
I considered `GroupBy`, though I was thinking it'd probably allocate more than a single `HashSet<string>` (and the underlying `Set<T>` in `Distinct()`). Probably want to group by a composite of `doc.Id` and `doc.Index`.

The `HashSet<string>` _could_ be initialized on line 38, but the allocation might be unneccessary if there are no ids.

Reply:
`GroupBy()` needs allocation or a comparer, I think the `HashSet<string>` although more imperative, is easier to read.